Google IAM

brunodomenici opened the pull request:
Hi,

We have an internal application as source of trust of ours Kafka ACLs and we implemented a `HTTP Source` in KSM with the ability to authenticate using `Google OpenID` implementation.
Also, we need to have Slack Notification and Console Notification, for this case we just added the log in Slack Notification too.

simplesteph commented on README.md:
```diff
@@ -189,6 +189,14 @@ The [default configurations](src/main/resources/application.conf) can be overwri
+    - `io.conduktor.ksm.source.HttpSourceAcl`: get the ACL from an HTTP endpoint. You can enable [Google OAuth OIDC Token Authentication](https://cloud.google.com/docs/authentication/production).
+        - `SOURCE_HTTP_URL` HTTP endpoint to retrieve ACL data.
+        - `SOURCE_HTTP_METHOD` HTTP Method. Default is `GET`.
+        - `SOURCE_HTTP_ENABLE_AUTH` If enable Google OICD Authentication. Default is `false`.
```
can we make the auth more generic in case more auth mechanisms are added in the future? (bearer etc...)

brunodomenici replied:
I thought about that, in fact, if we have a `pluggable` authentication method, I think you could even consolidate others HTTP sources like Github/Gitlab as they all are HTTP too. 
What you propose? An authentication class implementation for HTTP Source for example? How could we handle specifics configuration for this authentication class?

simplesteph commented on README.md:
```diff
@@ -189,6 +189,14 @@ The [default configurations](src/main/resources/application.conf) can be overwri
+        - `SOURCE_HTTP_URL` HTTP endpoint to retrieve ACL data.
+        - `SOURCE_HTTP_METHOD` HTTP Method. Default is `GET`.
+        - `SOURCE_HTTP_ENABLE_AUTH` If enable Google OICD Authentication. Default is `false`.
+        - `SOURCE_HTTP_SERVICE_ACCOUNT` Google Service Account name.
```
and therefore make these configs more specific to G OIDC as well

simplesteph reviewed (changes requested):
high level feedback. We'll review the code as well

simplesteph commented:
@trobert I think there should be a more "standard" way to create the Authentication to add other mechanisms, thoughts on how to organize this? can you review the code? 

@brunodomenici thanks for this. Can you explain what you mean by the notifications? They're separate from the Source ACL, and I don't see anything in your code related to Slack/Console, so I think you're just telling us about your usage patterns? 

Any ways to add tests as well?

brunodomenici commented:
> @brunodomenici thanks for this. Can you explain what you mean by the notifications? They're separate from the Source ACL, and I don't see anything in your code related to Slack/Console, so I think you're just telling us about your usage patterns?

@simplesteph sorry for mixing up subjects in this PR. But for the log/notification part, we do need to have both `Slack` + `Console` notification. Today if we use `io.conduktor.ksm.notification.SlackNotification` we don't have Console logs with KSM operations:

```
[2021-06-16 15:21:32,979] INFO ADDED User:247327, Topic:LITERAL:topic_test, Operation: Read, Permission: Allow, Host: * (io.conduktor.ksm.notification.SlackNotification)
[2021-06-16 15:21:43,181] INFO REMOVED User:test_bd, Topic:LITERAL:*, Operation: Describe, Permission: Allow, Host: * (io.conduktor.ksm.notification.SlackNotification)
[2021-06-16 15:21:43,182] INFO REMOVED User:test_bd, Topic:LITERAL:*, Operation: Read, Permission: Allow, Host: * (io.conduktor.ksm.notification.SlackNotification)
[2021-06-16 15:21:43,183] INFO REMOVED User:test_bd, Group:LITERAL:*, Operation: Read, Permission: Allow, Host: * (io.conduktor.ksm.notification.SlackNotification)
```

That's why I added a `log.info()` in `io.conduktor.ksm.notification.SlackNotification`

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,198 @@
+
+  private val log = LoggerFactory.getLogger(classOf[HttpSourceAcl])
+
+  private val jwtMap = collection.mutable.Map[String, String]()
+  private val expirationMap = collection.mutable.Map[String, Instant]()
```
why these maps ? the key is the audience but it seems the audience never change

brunodomenici replied:
You're no really needed

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,198 @@
+  private def getIdToken(): String = {
+    val audience = this.targetAudience
+
+    if (jwtMap.get(audience) != null
+      && expirationMap.get(audience) != null
```
`.get` returns an option, it can't be null

brunodomenici replied:
Done

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,198 @@
+      // Add cache for 30 minutes
+      log.info("Adding Token to cache for 30 minutes")
+      jwtMap.put(audience, jwt)
+      expirationMap.put(audience, Instant.now().plus(30, ChronoUnit.MINUTES))
```
can't we get the expiration from the token instead of a fixed 30 minutes ?

trobert replied:
at the minimum please make it configurable, if there is nothing at hand to parse the token

brunodomenici replied:
Using token expiration now...

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,198 @@
+        .createScoped(Collections.singleton(IAM_SCOPE))
+    }
+
+    if (!credentials.isInstanceOf[IdTokenProvider])
+      throw new RuntimeException(
+        s"Google Credentials: wrong type of token provider. Expected: IdTokenProvider, got: $credentials"
+      )
+
+    val idTokenCredentials = IdTokenCredentials.newBuilder
+      .setIdTokenProvider(credentials.asInstanceOf[ServiceAccountCredentials])
```
please use a `match` instead of `isInstanceOf` and check directly for the needed type (`ServiceAccountCredentials`)

brunodomenici replied:
Done!

trobert commented:
I agree we should separate Authentication in a separate trait / implementation. We can start with this one, and retrofit this to the other Source later.
for the config, we could have sth like:
```
SOURCE_HTTP_AUTH_TYPE="GoogleIAM" (or "None")
SOURCE_HTTP_AUTH_SERVICE_ACCOUNT=xxx
```

wdyt ?

trobert commented:
also I'd like to have some Test, especially around the token refresh part

brunodomenici commented:
Sorry for the delay. I really was with no time to work on this.

In this meantime, we had some issues regarding mismatch length between HTTP Header `Content-Length` and body length, so we also implemented an extra check.

Changed also the config to GOOGLE_IAM

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,216 @@
+    }
+  }
+
+  private def getIdToken(): String = {
```
we want to make addind new authentication easy. What do you think of isolating the authentication code behind an interface:
```scala
trait HttpAuthentication {
  def authHeader: String
}

class GoogleIAM(serviceAccountName: String, targetAudience: String, serviceAccountKey: String) extends HttpAuthentication {
```

and in `HttpSource` class:
```scala
  var authentication: Option[HttpAuthentication]

   authentication = config.getString(AUTHENTICATION_TYPE) match {
      case AuthenticationType.GOOGLE_IAM => Some(new GoogleIAM("TODO", "TODO", "TODO")) 
      case AuthenticationType.NONE => None
   }

```
(There should be no settings specific to google directly in the HttpSource class)

brunodomenici replied:
Good point. Let me know if the implementation is ok

trobert replied:
looks ok :+1:

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,216 @@
+    tokenCache.getRequestMetadata().get("Authorization").get(0) // Bearer Token
+  }
+
+  import java.io.IOException
+
+  @throws[IOException]
```
not necessary

brunodomenici replied:
👍

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,216 @@
+    credentials match {
+      case credentials: IdTokenProvider =>
+        IdTokenCredentials.newBuilder
+          .setIdTokenProvider(credentials.asInstanceOf[ServiceAccountCredentials])
```
why the cast? seems not necessary

brunodomenici replied:
Here we need to be sure to have an instance of `ServiceAccountCredentials` otherwise we can't use a credential key file

trobert replied:
so why not: `case credentials: ServiceAccountCredentials` ?

brunodomenici replied:
Sure! I think I need more practice in Scala 🤦🏻‍♂️

trobert commented on src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala:
```diff
@@ -0,0 +1,216 @@
+
+    val contentLengthHeader = optContentLengthHeader.get
+    val bodyLength = response.asBytes.length
+    log.info(s"Validating body length ($bodyLength bytes) received from $uri against Content-Length header ($contentLengthHeader bytes) claimed in response")
+    try {
+      val reader = new BufferedReader(new StringReader(response.textBody))
+      val aclLineCount = reader.lines.count
+      log.info("There were {} lines in the response received from {}", aclLineCount, uri)
+      reader.close()
+    }
+    catch {
+      case e: Exception => log.warn(s"Failed to compute number of lines in the response received from $uri", e)
+    }
```
logging body length and number of lines is not so useful to me

brunodomenici replied:
As we faced some errors in production regarding this, we included this on log to be sure if it was a problem in our backend or in skinny HTTP client. 
Can I change to trace level at least?

trobert replied:
fine. but this can be simplified:
```
    log.debug("There were {} lines in the response received from {}", response.textBody.lines().count(), uri)
```

trobert replied:
also can you add a comment that skinny http does not check content length so it  must be done manually, Thanks

brunodomenici replied:
Done

trobert reviewed (commented):
This is taking shape, thanks! 

Would you be able to look at the separation of the authentication method in a dedicated object? We don't want to be tied too much to google IAM. Thank you

brunodomenici commented on src/main/resources/application.conf:
```diff
@@ -134,6 +134,28 @@ source {
+    service_account_key = ${?SOURCE_HTTP_SERVICE_ACCOUNT_KEY}
+
+    target_audience = ""
+    target_audience = ${?SOURCE_HTTP_TARGET_AUDIENCE}
```
👍

trobert reviewed (approved):
LGTM !

trobert commented:
@simplesteph, can you approve the workflow and merge the PR if that's ok?

simplesteph commented:
CI fails

brunodomenici commented:
I see. I don't have the issue in local. In CI you're using Java 8, I'm using Java 11. I'll fix it...

brunodomenici commented:
@simplesteph , @trobert Java 8 combat fixed

simplesteph commented:
Thank you for your contribution!
